### PR TITLE
Add a basic feature flag system

### PR DESCRIPTION
**Pull request by evan2645**
This commit adds a basic system for feature flagging in SPIRE. As we
move towards a more continuous release process that allows us to safely
ship development features to users and contributors, feature flagging
will be a critical capability.

We have a long way to go - we still need to figure out how to deal with
feature flagging that affects our data models .. but for now, I think
this system should be good enough to get us most of the rest of the
cases.

This is just a first pass. Please feel free to pick it apart :)

Signed-off-by: Evan Gilman <egilman@vmware.com>

### CODE REVIEW

**Review by amartinezfayo — COMMENTED** [review]
Thank you for this @evan2645, it's looking great! I really like the simplicity of this solution, that at the same time gives us a powerful capability.

File: pkg/common/fflag/fflag.go
Comment (azdagron) [inline pkg/common/fflag/fflag.go:19]:
```suggestion
// false, with the only exception being flags that are in the process of being
```

File: pkg/common/fflag/fflag.go
Comment (azdagron) [inline pkg/common/fflag/fflag.go:51]: It seems to me that we'll need distinct feature flags for server and agent (that sometimes overlap). Seems to me that this  common code should just take a RawConfig and the flag definitions  and return a "feature flag set" that can be queried to enable server and agent do specify their own....

Comment (evan2645) [reply pkg/common/fflag/fflag.go:51]: > It seems to me that we'll need distinct feature flags for server and agent (that sometimes overlap).

I agree that we will have distinct features in agent or server that sometimes overlap, but I don't think that means that the feature flag namespace must also be disjoint. It feels nice to have one place where they are all defined .. if we allow server/agent to declare their own, we'll end up with two or possibly three places that a flag can be defined.

Do you think it's a bad idea to share the namespace like this? Even if only server or agent ever consumes a given flag, I don't think it ends up being problematic?

Comment (azdagron) [reply pkg/common/fflag/fflag.go:51]: It just means that you can accidentally set a flag on one that was intended for the other. You won't get error diagnostics that you goofed up, it will silently succeed but have any effect.

Comment (evan2645) [reply pkg/common/fflag/fflag.go:51]: Yea, that's not ideal ... but I also wonder how often this will actually bite someone - if you're trying to enable a flag that is specific to an agent or server (presumably because you are testing something), you're likely to quickly figure out that you're not working with the one you need?

IMO it is worth the trade-off, but I'm not opposed to splitting it up (especially if it's not the trade-off that others would make). I'd have to put some thought into how to do that ... I feel like I'd lean in the "other" direction (i.e. have some package-level flags defined in server + agent packages, and provide them from cli call site when calling the common Load).

Comment (azdagron) [reply pkg/common/fflag/fflag.go:51]: Evan and I chatted a bit. He's going to make a small change to the Load function that takes the server/agent specific flags so we can get good diagnostics.

**Comment by evan2645** [comment]
I pushed a few commits to hook fflag into the CLI, and to address CR feedback and linter.

As discussed in contributor sync today, I'm working on an additional commit that will split the flags into server, agent, and common. I'll send it as an individual commit so we can take it or leave it depending on how folks feel about it

**Comment by evan2645** [comment]
I pushed https://github.com/spiffe/spire/pull/2873/commits/6db912e7c9c74fab8ca06bd5da125c041ca58606 to add agent/server specific flags.

Please have a look and let me know what you all prefer. One thing we can do to minimize the pain is to move the flag default maps for agent and server into the common `fflag` package so you don't have to jump around as much, however I think it also is not how we have historically organized our code (e.g. the catalog).

File: pkg/common/fflag/fflag.go
Comment (evan2645) [inline pkg/common/fflag/fflag.go:79]: Should they?

Comment (azdagron) [reply pkg/common/fflag/fflag.go:79]: Hmm, I think taking common is probably not correct? I see us in one of two situations for shared feature flags:

1. Flag defaults must be consistent across server and agent. In this case, I would think it should be an error for the supplemental flag set to overlap with the common flag set. We should just have one flag defined.
2. Flag defaults can be different across server and agent. In this case, if we took common, there would be no way to override the default from the component. I think we'd want to take the supplemental default in this case to allow the server/agent to override the common default.

File: pkg/common/fflag/fflag.go
Comment (azdagron) [inline pkg/common/fflag/fflag.go:104]: nit: the validation code can be a lot more terse if it happens within `load()`, which can just delete flags from the `flagConfig` map as it applies them. Then, at the end, if there are any remaining flags in `flagConfig`, that would indicate an unknown flag. Doing so avoids this need of building up and mutating values in this badFlags set.

File: pkg/common/fflag/fflag.go
Comment (azdagron) [inline pkg/common/fflag/fflag.go:122]: nit: might be nice to sort the bad names for consistent error reporting between runs, which also has the side effect of making the error messages reproducible, which could help with deterministic testing of the resulting error message.

File: pkg/common/fflag/fflag_test.go
Comment (azdagron) [inline pkg/common/fflag/fflag_test.go:12]: super happy this is now scoped to just the test!!!

File: pkg/common/fflag/fflag_test.go
Comment (azdagron) [inline pkg/common/fflag/fflag_test.go:17]: nit: I think some of the test scaffolding (e.g. preSet and preUnset) could be removed and the testing simplified if we just defined two test flags; one that had a default true value, and the other a default false value.

File: pkg/agent/afflag/afflag.go
Comment (azdagron) [inline pkg/agent/afflag/afflag.go:8]: I don't have super strong opinions about this, but I feel like this could be an exception to our normal agent/server/common pattern and instead defined in pkg/common/fflag/agent.go. I don't see a huge benefit to keeping it separate.

If we do keep a separate package, I think we could unexport this map and provide a `Load(fflag.RawConfig) error` method in this package that turns around and calls `fflag.Load` with the now-unexported `agentFlagMap` :shrug: .

File: pkg/common/fflag/fflag.go
Comment (evan2645) [reply pkg/common/fflag/fflag.go:28]: Good call will change

File: pkg/common/fflag/fflag.go
Comment (azdagron) [reply pkg/common/fflag/fflag.go:79]: Ah gotcha. Yeah, I think an error makes the most sense :heart:

File: pkg/common/fflag/fflag.go
Comment (evan2645) [reply pkg/common/fflag/fflag.go:30]: Yes, I thought this would be ok because it is unexported.

I had some code that copied the commonFlagMap into this flags field as a new map, but it got a little messy and cumbersome and since commonFlagMap is not exported I thought why not just re-use it as flags field and mutate it... this was previously statically defined here (and mutated) but then the elements moved into a different map in `common.go` following the server/agent split

What is the advantage of copying the commonFlagMap elements into a new singleton.flags map?

File: pkg/common/fflag/fflag.go
Comment (evan2645) [reply pkg/common/fflag/fflag.go:56]: > nit: validateFlags and load are a little harder to reason about since they both touch global state (i.e. singleton.flags). I think the code would be easier to follow (and shorter) if they were combined and more functional

I'll have a look at this and see what I can do

> func load(config map[string]bool, flagSets ...map[Flag]bool) (map[Flag]bool, error)

I don't think we have any use cases for loading multiple flag sets beyond just common and agent/server? My preference on this work is to just build what we need right now

> Where the flagsSets are presented in descending priority order.

We can move to error condition so there is no priority

File: pkg/common/fflag/fflag.go
Comment (evan2645) [reply pkg/common/fflag/fflag.go:104]: 👍 nice I will look at that

File: pkg/common/fflag/fflag.go
Comment (evan2645) [reply pkg/common/fflag/fflag.go:122]: 👍 will do

File: pkg/common/fflag/fflag_test.go
Comment (evan2645) [reply pkg/common/fflag/fflag_test.go:12]: I'm not sure it can stay here 😭 I think I need to put it back .. still working through it ... we want to be able to assert that the statically defined common flags get loaded up, and to do that I think we need one defined there that we can test against. Since it is defined at the package level, I've had a hard time figuring out how to inject/manipulate it cleanly from the test suite

File: pkg/common/fflag/fflag_test.go
Comment (evan2645) [reply pkg/common/fflag/fflag_test.go:17]: nice! will do

File: pkg/common/fflag/fflag.go
Comment (azdagron) [reply pkg/common/fflag/fflag.go:56]: That suggested interface was meant to generalize just for convenience iterating over the flag sets inside the function, since the inner loop that processes each flag would be the same. I think having a distinct parameter for both supplemental and common is fine as well.

File: pkg/common/fflag/fflag_test.go
Comment (evan2645) [reply pkg/common/fflag/fflag_test.go:12]: > Since it is defined at the package level, I've had a hard time figuring out how to inject/manipulate it cleanly from the test suite

I'm specifically talking about the commonFlagMap var, not the flag itself

File: pkg/common/fflag/fflag.go
Comment (azdagron) [reply pkg/common/fflag/fflag.go:30]: To be clear, I'm not suggesting we copy it into singleton.flags here. I suggest commonFlagMap be passed or explicitly handled by `load()` to build the resulting flag set and `singleton.flags` be set from that returned value.

Also, I don't think anything is broken with the way it is currently done, but aliasing and unintentionally modifying this map that would otherwise likely be a constant if the language allowed for it, is just confusing and could potentially cause bugs in future refactors.

File: pkg/common/fflag/fflag_test.go
Comment (azdagron) [reply pkg/common/fflag/fflag_test.go:12]: Ah, for that, you can leverage an `init()` function in the `fflag_test.go` file to mutate the commonFlagMap just for tests without mutating it for production use.

```
func init() {
    commonFlagMap[flagTestFlag] = true
}
```

**Comment by evan2645** [comment]
> I don't have super strong opinions about this, but I feel like this could be an exception to our normal agent/server/common pattern and instead defined in pkg/common/fflag/agent.go. I don't see a huge benefit to keeping it separate.
>
> If we do keep a separate package, I think we could unexport this map and provide a Load(fflag.RawConfig) error method in this package that turns around and calls fflag.Load with the now-unexported agentFlagMap 🤷 (simplifying the imports in run/run.go).

Thanks @azdagron for the review 🤗 there are a handful of nits here that I'd like to pick up and push onto this PR .. however ..

It would be good to get some extra maintainer opinions on the agent/server/common split, and if/how we should accomplish it. My personal opinion is that splitting them up is more trouble than its worth. That being said, I also recognize that I'm effectively arguing for developer ergonomics over user experience, which is a hypocritical thing for me to do :). I guess my opinion is more nuanced than that, but I digress

I kept the split commit independent so we could decide which way we want to go, but for me to address the nits I need to build on top of something .. so it would be good if some of you could weigh in, especially @amartinezfayo who has already had a look at this PR in its earlier form

File: pkg/common/fflag/fflag_test.go
Comment (evan2645) [reply pkg/common/fflag/fflag_test.go:12]: oh neat! That's a good idea, I will try it out

File: pkg/common/fflag/fflag.go
Comment (evan2645) [reply pkg/common/fflag/fflag.go:30]: > To be clear, I'm not suggesting we copy it into singleton.flags here. I suggest commonFlagMap be passed or explicitly handled by load() to build the resulting flag set and singleton.flags be set from that returned value.

Gotcha .. this was/is difficult to do due to the current structure of the validate/load code and associated locking (which is why I opted for this shortcut), but with the help of a shakeup in Load (which will come by way of other CR comments you'e left), I think that is the right direction. Thank you!

> Also, I don't think anything is broken with the way it is currently done, but aliasing and unintentionally modifying this map that would otherwise likely be a constant if the language allowed for it, is just confusing and could potentially cause bugs in future refactors.

yes ,. it would have been a constant, and I would not have been allowed to do such trickeries! 😆

**Comment by evan2645** [comment]
> I kept the split commit independent so we could decide which way we want to go, but for me to address the nits I need to build on top of something .. so it would be good if some of you could weigh in, especially @amartinezfayo who has already had a look at this PR in its earlier form

Quick follow up here .. @azdagron and I have again discussed this offline. He feels more strongly than I do about making sure that the diagnostics are good, and that we not take a flag that doesn't have some effect. Of course, that fundamentally means that we must track which flags have an effect where.

At the same time, we recognized that perhaps we were not totally aligned on how these flags will be used and under what circumstances. We concluded that we should discuss this at our next maintainer sync (tomorrow), and that the outcome of this conversation should be recorded and added to this PR in the form of an amendment to the contribution guidelines.

**Comment by amartinezfayo** [comment]
Thanks for the update, @evan2645, I was going to comment about this.
I've been looking at the changes and specifically the commit that adds support to have agent/server specific flags. I have also been playing with how all this would work.
I'm leaning towards thinking that not being able to distinguish between agent and server flags can lead to a bad experience from the operator standpoint. And IMO, the developer experience is not that bad with the agent/server/common split.
I feel that even if the code gets a little more complex, we will gain a lot in terms of clarity for the operator, and avoid possible problems.
I'll happy to share my thoughts in the maintainer's sync.

**Comment by evan2645** [comment]
> I've been looking at the changes and specifically the commit that adds support to have agent/server specific flags. I have also been playing with how all this would work.
> I'm leaning towards thinking that not being able to distinguish between agent and server flags can lead to a bad experience from the operator standpoint. And IMO, the developer experience is not that bad with the agent/server/common split.
> I feel that even if the code gets a little more complex, we will gain a lot in terms of clarity for the operator, and avoid possible problems.

Awesome, thank you @amartinezfayo for this super valuable feedback. It aligns well with @azdagron's thoughts .. One thing that he and I discussed was the possibility of moving all server/agent/common flags declarations and maps into a single file in the common fflag package such that you can visit one place and see all flags, and where they apply. Consumers can then call fflag.AgentLoad or fflag.ServerLoad to get the set appropriate for them.

I will work on this improvement and a few others .. I will also address the outstanding nits. I still think we need to discuss the use of this facility, and add some text to the contributor's guidelines. Thanks again @amartinezfayo! It is enough to unblock me for now

**Comment by amartinezfayo** [comment]
One more thing that I forgot to comment about is that I think that it would be good to log when a feature flag is enabled. I don't think that this is being logged.

**Review by azdagron — APPROVED** [review]
\o/

**Pull request merged**
